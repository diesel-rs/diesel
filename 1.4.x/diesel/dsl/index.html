<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Includes various helper types and bare functions which are named too generically to be included in prelude, but are often used when using Diesel."><meta name="keywords" content="rust, rustlang, rust-lang, dsl"><title>diesel::dsl - Rust</title><link rel="stylesheet" type="text/css" href="../../normalize.css"><link rel="stylesheet" type="text/css" href="../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../../ayu.css" disabled ><script id="default-settings" ></script><script src="../../storage.js"></script><script src="../../crates.js"></script><noscript><link rel="stylesheet" href="../../noscript.css"></noscript><link rel="icon" type="image/svg+xml" href="../../favicon.svg"><link rel="alternate icon" type="image/png" href="../../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../../favicon-32x32.png"><style type="text/css">#crate-search{background-image:url("../../down-arrow.svg");}</style></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu" role="button">&#9776;</div><a href='../../diesel/index.html'><div class='logo-container rust-logo'><img src='../../rust-logo.png' alt='logo'></div></a><h2 class="location">Module dsl</h2><div class="sidebar-elems"><div class="block items"><ul><li><a href="#structs">Structs</a></li><li><a href="#functions">Functions</a></li><li><a href="#types">Type Definitions</a></li></ul></div><div id="sidebar-vars" data-name="dsl" data-ty="mod" data-relpath="./"></div><script defer src="./sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu" title="themes"><img width="18" height="18" alt="Pick another theme!" src="../../brush.svg"></button><div id="theme-choices" role="menu"></div></div><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input"name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><button type="button" id="help-button" title="help">?</button><a id="settings-menu" href="../../settings.html" title="settings"><img width="18" height="18" alt="Change settings" src="../../wheel.svg"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="in-band">Module <a href="../index.html">diesel</a>::<wbr><a class="mod" href="#">dsl</a><button id="copy-path" onclick="copy_path(this)" title="copy path"><img src="../../clipboard.svg" width="19" height="18" alt="Copy item import" title="Copy item import to clipboard"></button></span><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../../src/diesel/lib.rs.html#216-230" title="goto source code">[src]</a></span></h1><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Includes various helper types and bare functions which are named too
generically to be included in prelude, but are often used when using Diesel.</p>
</div></details><h2 id="reexports" class="section-header"><a href="#reexports">Re-exports</a></h2>
<div class="item-table"><div class="item-left import-item"><code>pub use pg::expression::<a class="mod" href="../pg/expression/dsl/index.html" title="mod diesel::pg::expression::dsl">dsl</a>::*;</code></div><div class="item-right docblock-short"></div></div><h2 id="structs" class="section-header"><a href="#structs">Structs</a></h2>
<div class="item-table"><div class="item-left module-item"><a class="struct" href="struct.now.html" title="diesel::dsl::now struct">now</a></div><div class="item-right docblock-short"><p>Represents the SQL <code>CURRENT_TIMESTAMP</code> constant. This is equivalent to the
<code>NOW()</code> function on backends that support it.</p>
</div></div><h2 id="functions" class="section-header"><a href="#functions">Functions</a></h2>
<div class="item-table"><div class="item-left module-item"><a class="fn" href="fn.avg.html" title="diesel::dsl::avg fn">avg</a></div><div class="item-right docblock-short"><p>Represents a SQL <code>AVG</code> function. This function can only take types which are
Foldable.</p>
</div><div class="item-left module-item"><a class="fn" href="fn.count.html" title="diesel::dsl::count fn">count</a></div><div class="item-right docblock-short"><p>Creates a SQL <code>COUNT</code> expression</p>
</div><div class="item-left module-item"><a class="fn" href="fn.count_star.html" title="diesel::dsl::count_star fn">count_star</a></div><div class="item-right docblock-short"><p>Creates a SQL <code>COUNT(*)</code> expression</p>
</div><div class="item-left module-item"><a class="fn" href="fn.date.html" title="diesel::dsl::date fn">date</a></div><div class="item-right docblock-short"><p>Represents the SQL <code>DATE</code> function. The argument should be a Timestamp
expression, and the return value will be an expression of type Date.</p>
</div><div class="item-left module-item"><a class="fn" href="fn.delete.html" title="diesel::dsl::delete fn">delete</a></div><div class="item-right docblock-short"><p>Creates a <code>DELETE</code> statement.</p>
</div><div class="item-left module-item"><a class="fn" href="fn.exists.html" title="diesel::dsl::exists fn">exists</a></div><div class="item-right docblock-short"><p>Creates a SQL <code>EXISTS</code> expression.</p>
</div><div class="item-left module-item"><a class="fn" href="fn.insert_into.html" title="diesel::dsl::insert_into fn">insert_into</a></div><div class="item-right docblock-short"><p>Creates an <code>INSERT</code> statement for the target table.</p>
</div><div class="item-left module-item"><a class="fn" href="fn.insert_or_ignore_into.html" title="diesel::dsl::insert_or_ignore_into fn">insert_or_ignore_into</a></div><div class="item-right docblock-short"><p>Creates an <code>INSERT [OR] IGNORE</code> statement.</p>
</div><div class="item-left module-item"><a class="fn" href="fn.max.html" title="diesel::dsl::max fn">max</a></div><div class="item-right docblock-short"><p>Represents a SQL <code>MAX</code> function. This function can only take types which are
ordered.</p>
</div><div class="item-left module-item"><a class="fn" href="fn.min.html" title="diesel::dsl::min fn">min</a></div><div class="item-right docblock-short"><p>Represents a SQL <code>MIN</code> function. This function can only take types which are
ordered.</p>
</div><div class="item-left module-item"><a class="fn" href="fn.not.html" title="diesel::dsl::not fn">not</a></div><div class="item-right docblock-short"><p>Creates a SQL <code>NOT</code> expression</p>
</div><div class="item-left module-item"><a class="fn" href="fn.replace_into.html" title="diesel::dsl::replace_into fn">replace_into</a></div><div class="item-right docblock-short"><p>Creates a <code>REPLACE</code> statement.</p>
</div><div class="item-left module-item"><a class="fn" href="fn.select.html" title="diesel::dsl::select fn">select</a></div><div class="item-right docblock-short"><p>Creates a bare select statement, with no from clause. Primarily used for
testing diesel itself, but likely useful for third party crates as well. The
given expressions must be selectable from anywhere.</p>
</div><div class="item-left module-item"><a class="fn" href="fn.sql.html" title="diesel::dsl::sql fn">sql</a></div><div class="item-right docblock-short"><p>Use literal SQL in the query builder</p>
</div><div class="item-left module-item"><a class="fn" href="fn.sql_query.html" title="diesel::dsl::sql_query fn">sql_query</a></div><div class="item-right docblock-short"><p>Construct a full SQL query using raw SQL.</p>
</div><div class="item-left module-item"><a class="fn" href="fn.sum.html" title="diesel::dsl::sum fn">sum</a></div><div class="item-right docblock-short"><p>Represents a SQL <code>SUM</code> function. This function can only take types which are
Foldable.</p>
</div><div class="item-left module-item"><a class="fn" href="fn.update.html" title="diesel::dsl::update fn">update</a></div><div class="item-right docblock-short"><p>Creates an <code>UPDATE</code> statement.</p>
</div></div><h2 id="types" class="section-header"><a href="#types">Type Definitions</a></h2>
<div class="item-table"><div class="item-left module-item"><a class="type" href="type.And.html" title="diesel::dsl::And type">And</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.BoolExpressionMethods.html#method.and"><code>lhs.and(rhs)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.AsExpr.html" title="diesel::dsl::AsExpr type">AsExpr</a></div><div class="item-right docblock-short"><p>The type of <code>Item</code> when converted to an expression with the same type as <code>TargetExpr</code></p>
</div><div class="item-left module-item"><a class="type" href="type.AsExprOf.html" title="diesel::dsl::AsExprOf type">AsExprOf</a></div><div class="item-right docblock-short"><p>The type of <code>Item</code> when converted to an expression of <code>Type</code></p>
</div><div class="item-left module-item"><a class="type" href="type.Asc.html" title="diesel::dsl::Asc type">Asc</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.ExpressionMethods.html#method.asc"><code>expr.asc()</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.Between.html" title="diesel::dsl::Between type">Between</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.ExpressionMethods.html#method.between"><code>lhs.between(lower, upper)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.Desc.html" title="diesel::dsl::Desc type">Desc</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.ExpressionMethods.html#method.desc"><code>expr.desc()</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.Distinct.html" title="diesel::dsl::Distinct type">Distinct</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.distinct()</code></p>
</div><div class="item-left module-item"><a class="type" href="type.DistinctOn.html" title="diesel::dsl::DistinctOn type">DistinctOn</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.distinct_on(expr)</code></p>
</div><div class="item-left module-item"><a class="type" href="type.Eq.html" title="diesel::dsl::Eq type">Eq</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.ExpressionMethods.html#method.eq"><code>lhs.eq(rhs)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.EqAny.html" title="diesel::dsl::EqAny type">EqAny</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.ExpressionMethods.html#method.eq_any"><code>lhs.eq_any(rhs)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.Escape.html" title="diesel::dsl::Escape type">Escape</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.EscapeExpressionMethods.html#method.escape"><code>lhs.escape('x')</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.Filter.html" title="diesel::dsl::Filter type">Filter</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.filter(predicate)</code></p>
</div><div class="item-left module-item"><a class="type" href="type.Find.html" title="diesel::dsl::Find type">Find</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.find(pk)</code></p>
</div><div class="item-left module-item"><a class="type" href="type.FindBy.html" title="diesel::dsl::FindBy type">FindBy</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.filter(lhs.eq(rhs))</code></p>
</div><div class="item-left module-item"><a class="type" href="type.ForKeyShare.html" title="diesel::dsl::ForKeyShare type">ForKeyShare</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.for_key_share()</code></p>
</div><div class="item-left module-item"><a class="type" href="type.ForNoKeyUpdate.html" title="diesel::dsl::ForNoKeyUpdate type">ForNoKeyUpdate</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.for_no_key_update()</code></p>
</div><div class="item-left module-item"><a class="type" href="type.ForShare.html" title="diesel::dsl::ForShare type">ForShare</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.for_share()</code></p>
</div><div class="item-left module-item"><a class="type" href="type.ForUpdate.html" title="diesel::dsl::ForUpdate type">ForUpdate</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.for_update()</code></p>
</div><div class="item-left module-item"><a class="type" href="type.Gt.html" title="diesel::dsl::Gt type">Gt</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.ExpressionMethods.html#method.gt"><code>lhs.gt(rhs)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.GtEq.html" title="diesel::dsl::GtEq type">GtEq</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.ExpressionMethods.html#method.ge"><code>lhs.ge(rhs)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.InnerJoin.html" title="diesel::dsl::InnerJoin type">InnerJoin</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.inner_join(rhs)</code></p>
</div><div class="item-left module-item"><a class="type" href="type.IntoBoxed.html" title="diesel::dsl::IntoBoxed type">IntoBoxed</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.into_boxed::&lt;'a, DB&gt;()</code></p>
</div><div class="item-left module-item"><a class="type" href="type.IsNotNull.html" title="diesel::dsl::IsNotNull type">IsNotNull</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.ExpressionMethods.html#method.is_not_null"><code>expr.is_not_null()</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.IsNull.html" title="diesel::dsl::IsNull type">IsNull</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.ExpressionMethods.html#method.is_null"><code>expr.is_null()</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.LeftJoin.html" title="diesel::dsl::LeftJoin type">LeftJoin</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.left_join(rhs)</code></p>
</div><div class="item-left module-item"><a class="type" href="type.Like.html" title="diesel::dsl::Like type">Like</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.TextExpressionMethods.html#method.like"><code>lhs.like(rhs)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.Limit.html" title="diesel::dsl::Limit type">Limit</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.limit()</code></p>
</div><div class="item-left module-item"><a class="type" href="type.Lt.html" title="diesel::dsl::Lt type">Lt</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.ExpressionMethods.html#method.lt"><code>lhs.lt(rhs)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.LtEq.html" title="diesel::dsl::LtEq type">LtEq</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.ExpressionMethods.html#method.le"><code>lhs.le(rhs)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.NeAny.html" title="diesel::dsl::NeAny type">NeAny</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.ExpressionMethods.html#method.ne_any"><code>lhs.ne_any(rhs)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.NoWait.html" title="diesel::dsl::NoWait type">NoWait</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.no_wait()</code></p>
</div><div class="item-left module-item"><a class="type" href="type.Not.html" title="diesel::dsl::Not type">Not</a><span class="stab deprecated" title="">Deprecated</span></div><div class="item-right docblock-short"><p>The return type of <code>not(expr)</code></p>
</div><div class="item-left module-item"><a class="type" href="type.NotBetween.html" title="diesel::dsl::NotBetween type">NotBetween</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.ExpressionMethods.html#method.not_between"><code>lhs.not_between(lower, upper)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.NotEq.html" title="diesel::dsl::NotEq type">NotEq</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.ExpressionMethods.html#method.ne"><code>lhs.ne(rhs)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.NotLike.html" title="diesel::dsl::NotLike type">NotLike</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.TextExpressionMethods.html#method.not_like"><code>lhs.not_like(rhs)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.Nullable.html" title="diesel::dsl::Nullable type">Nullable</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.NullableExpressionMethods.html#method.nullable"><code>expr.nullable()</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.NullableSelect.html" title="diesel::dsl::NullableSelect type">NullableSelect</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.nullable()</code></p>
</div><div class="item-left module-item"><a class="type" href="type.Offset.html" title="diesel::dsl::Offset type">Offset</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.offset()</code></p>
</div><div class="item-left module-item"><a class="type" href="type.Or.html" title="diesel::dsl::Or type">Or</a></div><div class="item-right docblock-short"><p>The return type of
<a href="../expression_methods/trait.BoolExpressionMethods.html#method.or"><code>lhs.or(rhs)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.OrFilter.html" title="diesel::dsl::OrFilter type">OrFilter</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.or_filter(predicate)</code></p>
</div><div class="item-left module-item"><a class="type" href="type.Order.html" title="diesel::dsl::Order type">Order</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.order(ordering)</code></p>
</div><div class="item-left module-item"><a class="type" href="type.Select.html" title="diesel::dsl::Select type">Select</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.select(selection)</code></p>
</div><div class="item-left module-item"><a class="type" href="type.SingleValue.html" title="diesel::dsl::SingleValue type">SingleValue</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.single_value()</code></p>
</div><div class="item-left module-item"><a class="type" href="type.SkipLocked.html" title="diesel::dsl::SkipLocked type">SkipLocked</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.skip_locked()</code></p>
</div><div class="item-left module-item"><a class="type" href="type.SqlTypeOf.html" title="diesel::dsl::SqlTypeOf type">SqlTypeOf</a></div><div class="item-right docblock-short"><p>The SQL type of an expression</p>
</div><div class="item-left module-item"><a class="type" href="type.ThenOrderBy.html" title="diesel::dsl::ThenOrderBy type">ThenOrderBy</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>.then_order_by(ordering)</code></p>
</div><div class="item-left module-item"><a class="type" href="type.Update.html" title="diesel::dsl::Update type">Update</a></div><div class="item-right docblock-short"><p>Represents the return type of <code>update(lhs).set(rhs)</code></p>
</div><div class="item-left module-item"><a class="type" href="type.avg.html" title="diesel::dsl::avg type">avg</a></div><div class="item-right docblock-short"><p>The return type of <a href="../dsl/fn.avg.html"><code>avg(expr)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.count.html" title="diesel::dsl::count type">count</a></div><div class="item-right docblock-short"><p>The return type of <a href="../dsl/fn.count.html"><code>count(expr)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.count_star.html" title="diesel::dsl::count_star type">count_star</a></div><div class="item-right docblock-short"><p>The return type of <a href="../dsl/fn.count_star.html"><code>count_star)(</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.date.html" title="diesel::dsl::date type">date</a></div><div class="item-right docblock-short"><p>The return type of <a href="../dsl/fn.date.html"><code>date(expr)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.max.html" title="diesel::dsl::max type">max</a></div><div class="item-right docblock-short"><p>The return type of <a href="../dsl/fn.max.html"><code>max(expr)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.min.html" title="diesel::dsl::min type">min</a></div><div class="item-right docblock-short"><p>The return type of <a href="../dsl/fn.min.html"><code>min(expr)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.not.html" title="diesel::dsl::not type">not</a></div><div class="item-right docblock-short"><p>The return type of <a href="../dsl/fn.not.html"><code>not(expr)</code></a></p>
</div><div class="item-left module-item"><a class="type" href="type.sum.html" title="diesel::dsl::sum type">sum</a></div><div class="item-right docblock-short"><p>The return type of <a href="../dsl/fn.sum.html"><code>sum(expr)</code></a></p>
</div></div></section><section id="search" class="content hidden"></section><div id="rustdoc-vars" data-root-path="../../" data-current-crate="diesel" data-search-index-js="../../search-index.js" data-search-js="../../search.js"></div>
    <script src="../../main.js"></script>
</body></html>