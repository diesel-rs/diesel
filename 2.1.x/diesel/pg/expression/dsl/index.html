<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="PostgreSQL specific expression DSL methods."><title>diesel::pg::expression::dsl - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../../../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../../../../static.files/rustdoc-081576b923113409.css"><meta name="rustdoc-vars" data-root-path="../../../../" data-static-root-path="../../../../static.files/" data-current-crate="diesel" data-themes="" data-resource-suffix="" data-rustdoc-version="1.79.0-nightly (e3181b091 2024-04-18)" data-channel="nightly" data-search-js="search-5b46ae87cc198d72.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../../../../static.files/storage-e32f0c247825364d.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../../../static.files/main-20a3ad099b048cf2.js"></script><noscript><link rel="stylesheet" href="../../../../static.files/noscript-09095024cf37855e.css"></noscript><link rel="alternate icon" type="image/png" href="../../../../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../../../../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../../../diesel/index.html">diesel</a><span class="version">2.1.6</span></h2></div><h2 class="location"><a href="#">Module dsl</a></h2><div class="sidebar-elems"><section><ul class="block"><li><a href="#traits">Traits</a></li><li><a href="#functions">Functions</a></li></ul></section><h2><a href="../index.html">In diesel::pg::expression</a></h2></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><div id="sidebar-button" tabindex="-1"><a href="../../../../diesel/all.html" title="show sidebar"></a></div><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Type ‘S’ or ‘/’ to search, ‘?’ for more options…" type="search"><div id="help-button" tabindex="-1"><a href="../../../../help.html" title="help">?</a></div><div id="settings-menu" tabindex="-1"><a href="../../../../settings.html" title="settings">Settings</a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Module <a href="../../../index.html">diesel</a>::<wbr><a href="../../index.html">pg</a>::<wbr><a href="../index.html">expression</a>::<wbr><a class="mod" href="#">dsl</a><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><span class="out-of-band"><a class="src" href="../../../../src/diesel/pg/expression/mod.rs.html#23">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><span class="item-info"><div class="stab portability">Available on <strong>crate feature <code>postgres_backend</code></strong> only.</div></span><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>PostgreSQL specific expression DSL methods.</p>
<p>This module will be glob imported by
<a href="../../../dsl/index.html" title="mod diesel::dsl"><code>diesel::dsl</code></a> when compiled with the <code>feature = &quot;postgres&quot;</code> flag.</p>
</div></details><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="trait" href="trait.IntervalDsl.html" title="trait diesel::pg::expression::dsl::IntervalDsl">IntervalDsl</a></div><div class="desc docblock-short">A DSL added to integers and <code>f64</code> to construct PostgreSQL intervals.</div></li><li><div class="item-name"><a class="trait" href="trait.OnlyDsl.html" title="trait diesel::pg::expression::dsl::OnlyDsl">OnlyDsl</a></div><div class="desc docblock-short">The <code>only</code> method</div></li></ul><h2 id="functions" class="section-header">Functions<a href="#functions" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="fn" href="fn.abbrev.html" title="fn diesel::pg::expression::dsl::abbrev">abbrev</a></div><div class="desc docblock-short">Creates an abbreviated display format as text.</div></li><li><div class="item-name"><a class="fn" href="fn.all.html" title="fn diesel::pg::expression::dsl::all">all</a><span class="stab deprecated" title="">Deprecated</span><span class="stab portability" title="Available on crate feature `with-deprecated` and non-crate feature `without-deprecated` only"><code>with-deprecated</code> and non-<code>without-deprecated</code></span></div><div class="desc docblock-short">Creates a PostgreSQL <code>ALL</code> expression.</div></li><li><div class="item-name"><a class="fn" href="fn.any.html" title="fn diesel::pg::expression::dsl::any">any</a><span class="stab deprecated" title="">Deprecated</span><span class="stab portability" title="Available on crate feature `with-deprecated` and non-crate feature `without-deprecated` only"><code>with-deprecated</code> and non-<code>without-deprecated</code></span></div><div class="desc docblock-short">Creates a PostgreSQL <code>ANY</code> expression.</div></li><li><div class="item-name"><a class="fn" href="fn.array.html" title="fn diesel::pg::expression::dsl::array">array</a></div><div class="desc docblock-short">Creates an <code>ARRAY[...]</code> expression.</div></li><li><div class="item-name"><a class="fn" href="fn.broadcast.html" title="fn diesel::pg::expression::dsl::broadcast">broadcast</a></div><div class="desc docblock-short">Computes the broadcast address for the address’s network.</div></li><li><div class="item-name"><a class="fn" href="fn.family.html" title="fn diesel::pg::expression::dsl::family">family</a></div><div class="desc docblock-short">Returns the address’s family: 4 for IPv4, 6 for IPv6.</div></li><li><div class="item-name"><a class="fn" href="fn.host.html" title="fn diesel::pg::expression::dsl::host">host</a></div><div class="desc docblock-short">Returns the IP address as text, ignoring the netmask.</div></li><li><div class="item-name"><a class="fn" href="fn.hostmask.html" title="fn diesel::pg::expression::dsl::hostmask">hostmask</a></div><div class="desc docblock-short">Computes the host mask for the address’s network.</div></li><li><div class="item-name"><a class="fn" href="fn.inet_merge.html" title="fn diesel::pg::expression::dsl::inet_merge">inet_merge</a></div><div class="desc docblock-short">Computes the smallest network that includes both of the given networks.</div></li><li><div class="item-name"><a class="fn" href="fn.inet_same_family.html" title="fn diesel::pg::expression::dsl::inet_same_family">inet_same_family</a></div><div class="desc docblock-short">Tests whether the addresses belong to the same IP family.</div></li><li><div class="item-name"><a class="fn" href="fn.masklen.html" title="fn diesel::pg::expression::dsl::masklen">masklen</a></div><div class="desc docblock-short">Returns the netmask length in bits.</div></li><li><div class="item-name"><a class="fn" href="fn.netmask.html" title="fn diesel::pg::expression::dsl::netmask">netmask</a></div><div class="desc docblock-short">Computes the network mask for the address’s network.</div></li><li><div class="item-name"><a class="fn" href="fn.network.html" title="fn diesel::pg::expression::dsl::network">network</a></div><div class="desc docblock-short">Returns the network part of the address, zeroing out whatever is to the right of the
netmask. (This is equivalent to casting the value to cidr.)</div></li><li><div class="item-name"><a class="fn" href="fn.set_masklen.html" title="fn diesel::pg::expression::dsl::set_masklen">set_masklen</a></div><div class="desc docblock-short">Sets the netmask length for an inet or cidr value.
For inet, the address part does not changes. For cidr, address bits to the right of the new
netmask are set to zero.</div></li></ul></section></div></main></body></html>