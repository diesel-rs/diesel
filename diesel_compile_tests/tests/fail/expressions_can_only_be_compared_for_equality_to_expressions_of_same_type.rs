extern crate diesel;

use diesel::*;

table! {
    users {
        id -> Integer,
        name -> VarChar,
    }
}

fn main() {
    use self::users::dsl::*;

    let pred = id.eq("string");
    //~^ ERROR: the trait bound `str: diesel::Expression` is not satisfied
    let pred = id.eq(name);
    //~^ ERROR: type mismatch resolving `<name as Expression>::SqlType == Integer`
}
