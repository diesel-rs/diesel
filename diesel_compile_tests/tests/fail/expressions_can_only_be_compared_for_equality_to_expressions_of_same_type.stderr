error[E0277]: the trait bound `str: diesel::Expression` is not satisfied
  --> $DIR/expressions_can_only_be_compared_for_equality_to_expressions_of_same_type.rs:15:22
   |
15 |     let pred = id.eq("string");
   |                      ^^^^^^^^ the trait `diesel::Expression` is not implemented for `str`
   |
   = note: required because of the requirements on the impl of `diesel::Expression` for `&str`
   = note: required because of the requirements on the impl of `AsExpression<diesel::sql_types::Integer>` for `&str`

error[E0271]: type mismatch resolving `<columns::name as diesel::Expression>::SqlType == diesel::sql_types::Integer`
  --> $DIR/expressions_can_only_be_compared_for_equality_to_expressions_of_same_type.rs:16:19
   |
16 |     let pred = id.eq(name);
   |                   ^^ expected struct `diesel::sql_types::Text`, found struct `diesel::sql_types::Integer`
   |
   = note: required because of the requirements on the impl of `AsExpression<diesel::sql_types::Integer>` for `columns::name`
